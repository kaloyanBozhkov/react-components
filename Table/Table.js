import React from 'react'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'

import styles from './styles.module.scss'

//import Pagination from '~components/Pagination/Pagination'

import Icon from 'UI/Icon/Icon'

//import helpers
import dateDisplay from '~/utilities/dateDisplay/dateDisplay'

//create table header from name of properties of rows returned from API
const createHeader = (columnHeaders = []) => (
    <header className={styles.tableHeader}>
        {columnHeaders.map(({ label, centered = false, withIcon = false, onClick = undefined }, index) => {

            //apply specific styling based on needs for that table
            const classes = [
                centered ? styles.centered : '',
                withIcon ? styles.withIcon : '',
                onClick ? styles.withAction : ''
            ].join(' ').trim() || undefined

            return (
                <div key={index} className={classes} onClick={onClick}>
                    <p>{label}</p>{withIcon && <Icon icon={withIcon} />}
                </div>
            )
        })}
    </header>
)

//create table body rows
    const createRows = (rows = []) => (
        <section className={styles.tableBody} emptymessage={rows.length === 0? "No data to show yet, try again later!" : undefined} >
            {rows.map((row, index) => <div key={index} className={styles.tableRow}>{
                Object.values(row).map(({ cellValue, cellType, centered = false, onClick = null }, index) => {
                   
                    const centeredCellContents = centered ? styles.centeredCellContents : undefined
                    const onClickFunc = onClick ? (() => onClick({ cellValue, rowIndex: index })) : undefined
                   
                    switch (cellType) {
                        case 'text':
                            return (
                                <div key={index} className={centeredCellContents} onClick={onClickFunc} >
                                    <p>{cellValue}</p>
                                </div>
                            )
                        case 'icon':
                            return (
                                <div icon={cellValue} key={index} onClick={onClickFunc} >
                                    <Icon icon={cellValue} className={centeredCellContents} />
                                </div>
                            )
                        case 'link':
                            if (typeof cellValue === 'object' && cellValue.hasOwnProperty('path') && cellValue.hasOwnProperty('label')) {
                                return (
                                    <div key={index} className={styles.urlDiv} onClick={onClickFunc}>
                                        <Link className={centeredCellContents} to={cellValue.path}>{cellValue.label}</Link>
                                    </div>
                                )
                            }

                            //or return a UI friendly msg?
                            return null                    
                        case 'date':
                            const parseDateStringOrDate = new Date(cellValue)

                            //check if invalid date obj
                            if (isNaN(parseDateStringOrDate.getTime())) {

                                //or return a UI friendly msg?
                                return null
                            }

                            return (
                                <div key={index} className={styles.dateDiv + ' ' + centeredCellContents} onClick={onClickFunc}>
                                    <p><b>{dateDisplay(parseDateStringOrDate)}</b></p>
                                    <p>{parseDateStringOrDate.toLocaleTimeString()}</p>
                                </div>
                            )
                        // case 'textTwoParts': //same formatting as date, just sets normal text
                        //     return (
                        //         <div key={index} className={styles.dateDiv + ' ' + centeredCellContents} onClick={onClickFunc}>
                        //             <p><b>{cellValue.first}</b></p>
                        //             <p>{cellValue.second}</p>
                        //         </div>
                        //     )
                        // case 'lotteryNumbers'://not yet implemented/tested since no img?
                        //     return (
                        //         <div key={index} className={centeredCellContents} onClick={onClickFunc} >
                        //             {/* <LotteryNumbers label={cellValue.label} state={cellValue.state} numbersArr={cellValue.numbers} /> */}
                        //         </div>
                        //     )
                        // case 'img'://not yet implemented/tested since no img?
                        //     return (
                        //         <img src={cellValue} alt={cellValue} onClick={onClickFunc} />
                        //     )
                        default:
                            return (
                                <div key={index} className={centeredCellContents} onClick={onClickFunc} >
                                    <p>-</p>
                                </div>
                            )
                    }

                })}</div>)}
        </section>
    )

/**
 * @param  {} rows -> aar of obj: {  cellValue, cellType, centered?, onClick? }
 * @param  {} columnHeaders -> arr of obj: { label, centered? , withIcon?, onClick? }
 * @param  {} noActions -> bool, do the table rows have any Active/Edit/Delete cells to click?
 * @param  {} pageName -> string, unique string used in Table component's styles module scss to tell it the page specific grid-columns-template 
 */
const Table = ({ pageName, columnHeaders = [], rows = [], noActions = false }) => {

    //show a user friendly msg if provided columnHeaders are missing, or throw Error?
    if (!columnHeaders.length) {
        return <p>Could not display table, please try again later</p>
    }

    const classes = [
        styles.table,
        noActions ? styles.noActions : '',
        styles[pageName] || ''
    ].join(' ').trim()

    return (
        <div className={classes}>
            {createHeader(columnHeaders)}
            {createRows(rows)}
        </div>
    )
}

Table.propTypes = {
     rows: PropTypes.arrayOf(
        PropTypes.arrayOf(
            PropTypes.shape({
                cellValue: PropTypes.oneOfType([
                    //label/text
                    PropTypes.string,
                    //timestamp
                    PropTypes.number, 
                    //link obj
                    PropTypes.shape({
                        path: PropTypes.string,
                        label: PropTypes.string
                    }),
                    //other proptypes based on added cell types for row generation
                ]).isRequired, 
                cellType: PropTypes.oneOf(['text', 'icon', 'link', 'date', 'img']), 
                centered: PropTypes.bool, 
                onClick: PropTypes.func
            })
        )
     ), 
     columnHeaders: PropTypes.oneOfType([
         PropTypes.arrayOf([
            PropTypes.shape({
                label: PropTypes.string.isRequired, 
                centered: PropTypes.bool, 
                withIcon: PropTypes.string
            })
        ]),
        PropTypes.array
     ]).isRequired, 
     noActions: PropTypes.bool,
     pageName: PropTypes.string.isRequired
}

export default Table
