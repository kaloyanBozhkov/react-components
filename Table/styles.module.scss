@import "scss/abstracts/_mixins";
@import "scss/abstracts/_variables";

.table {
  font-weight: bold;
  color: $primary-text;
  overflow: auto;

  //shadow for uverflow case, to have border left-right showing
  @include boxShadow(false, 0, 0, 7px, -5px, 0.35);

  &:not(.noActions) {
    .tableRow > div {
      &[icon="edit"],
      &[icon="trash-alt"],
      &[icon="trash"] {
        cursor: pointer;
        &:hover {
          svg {
            transform: scale(0.85);
          }
        }
      }

      &[icon="edit"] {
        color: $purple;
      }

      &[icon="check"] {
        color: $success-color;
      }
    }
  }

  .tableHeader {
    grid-gap: 1px;
    height: 4.2rem;
    background-color: #eaecef;
    padding: 1px 1px 0px 1px;

    > div {
      display: flex;
      align-items: center;

      font-size: 1rem;
      text-transform: uppercase;

      transition: 300ms;
      background-color: #f2f5f7;

      &.withAction {
        cursor:pointer;
        &:hover {
          background-color: #e5e5e9;
        }
      }

      padding: $medium-padding * 1.5;

      &.centered {
        justify-content: center;
      }

      &.withIcon {
        p {
          padding-right: 2rem;
        }
      }
    }
  }

  .tableBody {
    flex-grow: 1;

    &:empty {
      &::before {
        content: attr(emptymessage);
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        height: 15rem;
        background-color: #eaecef;
        @include boxShadow(false, 0, 2px, 2px, 1px, 0.1);
      }
    }

    .tableRow {
      grid-gap: 1px;

      background-color: #eaecef;

      &:first-of-type {
        padding: 1px;
      }

      &:not(:first-of-type) {
        padding: 0px 1px 1px 1px;
      }

      &:hover {
        > div {
          background-color: #f5f5f5;
        }
      }

      > div {
        display: flex;
        align-items: center;

        font-size: 1.2rem;
        transition: 200ms;
        background-color: white;

        padding: $medium-padding * 1.5;

        &.centeredCellContents {
          justify-content: center;
        }

        &.urlDiv {
          a {
            color: $secondary-color;
            text-decoration: none;
            &:hover {
              color: darken($secondary-color, 15%);
            }
          }
        }

        &.dateDiv {
          flex-direction: column;
          &.cellWrapperContentsLeft {
            align-items: flex-start;
          }
          p {
            &:first-of-type {
              font-weight: bold;
            }
            &:last-of-type {
              font-weight: normal;
            }
          }
        }

        &[icon] {
          //icons and images center here!!
          justify-content: center;
        }
      }
    }
  }

  .tableHeader,
  .tableBody {
    min-width: 600px;

    > div {
      &:nth-child(2),
      &:nth-child(3) {
        min-width: 5rem;
      }
    }
  }

  .tableHeader,
  .tableRow {
    display: grid;
  }

  //CUSTOM STYLINGS BASED ON CONTENT HERE
  &.games {
    .tableHeader,
    .tableRow {
      grid-template-columns: repeat(2, 8.5rem) 1fr 0.67fr repeat(3, 8.5rem);
    }
    .tableRow {
      &.games {
        grid-template-rows: minmax(7.5rem, auto);
      }
    }
  }
  &.gameFilters {
    .tableHeader,
    .tableRow {
      grid-template-columns: repeat(2, 1fr) repeat(2, 8.5rem);
    }
    .tableRow {
      &.games {
        grid-template-rows: auto;
      }
    }
  }
  &.components {
    .tableHeader,
    .tableRow {
      grid-template-columns: repeat(2, 1fr) repeat(3, 8.5rem);
    }
    .tableRow {
      &.games {
        grid-template-rows: auto;
      }
    }
  }
}
