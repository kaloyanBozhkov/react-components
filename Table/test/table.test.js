import React from 'react'
import { shallow, mount } from 'enzyme'

import Table from 'Collections/Table/Table'

describe('testing Table component rendering', () => {
    it('should render table component not configured properly msg, without any props ', () => {
        const wrapper = shallow(<Table columnHeaders={[]} pageName="someName" />)

        expect(wrapper).toMatchSnapshot()
    })


    //maybe we dont need this second snapshot test?
    it('should render table component when configured properly ', () => {
        const headers = [
            {
                label: 'columnHeader'
            }
        ]
        const wrapper = shallow(<Table  columnHeaders={headers} pageName="someName"  />)

        expect(wrapper).toMatchSnapshot()
    })
})

describe('testing Table component conditional renders for header', () => {
    it('should render multiple columns/headers with proper label values', () => {
        const headers = [
            {
                label: 'columnHeader1'
            },
            {
                label: 'columnHeader2',
            },            
            {
                label: 'columnHeader3',
            },
        ]
        
        const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" />)

        //3 columns/headers
        expect(wrapper.find('.tableHeader').children()).toHaveLength(3)

        //test label content
        expect(wrapper.find('.tableHeader').childAt(0).find('p').text()).toEqual('columnHeader1')
        expect(wrapper.find('.tableHeader').childAt(1).find('p').text()).toEqual('columnHeader2')
        expect(wrapper.find('.tableHeader').childAt(2).find('p').text()).toEqual('columnHeader3')
    })

    it('should render just a header with label, no optional classes', () => {
        const headers = [
            {
                label: 'columnHeader'
            }
        ]
        
        const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" />)
        const headerCell = wrapper.find('.tableHeader').childAt(0)

        //test their unique display related classes
        expect(headerCell.hasClass('centered')).toEqual(false)
        expect(headerCell.hasClass('withIcon')).toEqual(false)
    })

    it('should render just a header with label, and optional class "centered"', () => {
        const headers = [
            {
                label: 'columnHeader',
                centered: true
            }
        ]
        
        const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" />)
        const centeredHeaderCell = wrapper.find('.tableHeader').childAt(0)

        expect(centeredHeaderCell.hasClass('centered')).toEqual(true)
        expect(centeredHeaderCell.hasClass('withIcon')).toEqual(false)
    })

    it('should render just a header with label, and optional class "withIcon"', () => {
        const headers = [
            {
                label: 'columnHeader',
                withIcon: true
            }
        ]
        
        const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" />)
        const withIconHeaderCell = wrapper.find('.tableHeader').childAt(0)

        expect(withIconHeaderCell.hasClass('centered')).toEqual(false)
        expect(withIconHeaderCell.hasClass('withIcon')).toEqual(true)
        expect(withIconHeaderCell.find('Icon')).toHaveLength(1)
    })

    it('should render just a header with label, and optional class "withIcon" and "centered', () => {
        const headers = [
            {
                label: 'columnHeader',
                centered: true,
                withIcon: 'sort'
            },    
        ]
        
        const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" />)
        const headerCell = wrapper.find('.tableHeader').childAt(0)

        expect(headerCell.hasClass('centered')).toEqual(true)
        expect(headerCell.hasClass('withIcon')).toEqual(true)
    })

    it('should render jsut a header with label, that is clickable and can run func', () => {
        const mockFunc = jest.fn(() => "works")

        const headers = [
            {
                label: 'columnHeader',
                onClick: mockFunc
            }
        ]
        
        const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" />)
        const headerCell = wrapper.find('.tableHeader').childAt(0)
        headerCell.simulate('click')
        
        expect(mockFunc).toHaveBeenCalled()
    })
})

describe('testing Table component rows', () => {
    describe('testing row rendering', () => {

        it('should render table component without rows', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" />)
            

            //row container should have 'emptymessage' attribute which styles contents when no rows to show
            expect(wrapper.find('.tableBody[emptymessage]')).toHaveLength(1)
            expect(wrapper.find('.tableRow')).toHaveLength(0)
        })

        it('should render table component with a single row & column, with default cell value', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: 'hello world'
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)

            expect(wrapper.find('.tableBody[emptymessage]')).toHaveLength(0)
            expect(wrapper.find('.tableRow')).toHaveLength(1)

            //count rows created
            expect(wrapper.find('.tableBody').children()).toHaveLength(1)

            //test default label when misconfigured cell
            expect(wrapper.find('.tableBody').childAt(0).text()).toEqual('-')
        })

        it('should render table component with 2 rows, with default cell value', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: 'hello world'
                    }
                ],
                [
                    {
                        cellValue: 'hello world'
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)

            //count rows created
            expect(wrapper.find('.tableRow')).toHaveLength(2)

            expect(wrapper.find('.tableBody').childAt(0).find('p').text()).toEqual('-')
            expect(wrapper.find('.tableBody').childAt(1).find('p').text()).toEqual('-')
        })
    })
    
    describe('testing for text type cells', () => {
        
         it('should render table component with a single row & column, with text cell type', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: 'hello world',
                        cellType: 'text'
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)

            //test default label when configured cell
            expect(wrapper.find('.tableBody').childAt(0).text()).toEqual('hello world')
        })       

        it('should render table component with a single row & column, with text cell type wiht optional class', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: 'hello world',
                        cellType: 'text',
                        centered: true
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)
            const rowCell = wrapper.find('.tableRow').childAt(0)
            
            //test default label when configured cell
            expect(rowCell.hasClass('centeredCellContents')).toEqual(true)
        })

        it('should render table component with a single row & column, with text cell type wiht optional click function', () => {
            
            const mockFunc = jest.fn(() => "works")

            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: 'hello world',
                        cellType: 'text',
                        onClick: mockFunc
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)
            const rowCell = wrapper.find('.tableRow').childAt(0)
            rowCell.simulate('click')

            //test default label when configured cell
            expect(mockFunc).toHaveBeenCalled()
        })
    })

    describe('renderign table rows, with Icon cell type', () => {

        //do tests for click, opt classes? They are the same as text, just contents change. If we should, put them here! Otherwise, remove the describe block?

        it('should render table component with a single row & column, with Icon cell type', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: 'check',
                        cellType: 'icon'
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)

            expect(wrapper.find('.tableRow > div[icon="check"]')).toHaveLength(1)
            expect(wrapper.find('.tableRow').text()).toEqual('<Icon />')
        })

    })

    describe('testing table row rendering link type', () => {

        it('should render table component with a single row & column, with link cell type', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: {
                            path: 'someRoute',
                            label: 'redirectMe'
                        },
                        cellType: 'link'
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)
            const rowCell = wrapper.find('.tableRow').childAt(0)
            
            expect(rowCell.hasClass('urlDiv')).toEqual(true)
        })

        it('should handle wrong obj format for link', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: {
                            wrong: 'redirectMe'
                        },
                        cellType: 'link'
                    }
                ]
            ]

            //@TO DO: handle wrong prop types?! https://medium.com/shark-bytes/type-checking-with-prop-types-in-jest-e0cd0dc92d5
            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)

            //dont render Link if 
            expect(wrapper.find('.tableRow').children()).toHaveLength(0)
        })

    })
    
    describe('test row renderign with date cells', () => {

        it('should render table component with a single row & column, with date cell type', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: new Date('1990'),
                        cellType: 'date'
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)
            const rowCell = wrapper.find('.tableRow').childAt(0)
            
            expect(rowCell.hasClass('dateDiv')).toEqual(true)
            expect(rowCell.children()).toHaveLength(2)
            expect(rowCell.childAt(0).childAt(0).text()).toEqual('1st Jan 1990')
        })

        it('should render table component with a single row & column, with date cell type and string date', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: '1/1/1990',
                        cellType: 'date'
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)
            const rowCell = wrapper.find('.tableRow').childAt(0)
            
            expect(rowCell.hasClass('dateDiv')).toEqual(true)
            expect(rowCell.children()).toHaveLength(2)
            expect(rowCell.childAt(0).childAt(0).text()).toEqual('1st Jan 1990')
        })

        it('should render table component with a single row & column, with date cell type and timestamp date', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: 631152000000,
                        cellType: 'date'
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)
            const rowCell = wrapper.find('.tableRow').childAt(0)
            
            expect(rowCell.hasClass('dateDiv')).toEqual(true)
            expect(rowCell.children()).toHaveLength(2)
            expect(rowCell.childAt(0).childAt(0).text()).toEqual('1st Jan 1990')
        })

        it('should render table component with a single row & column, with date cell type and fail due to invalid date', () => {
            const headers = [
                {
                    label: 'columnHeader'
                }
            ]

            //arr of rows
            const rows = [
                [
                    {
                        cellValue: 'asd',
                        cellType: 'date'
                    }
                ]
            ]

            const wrapper = shallow(<Table columnHeaders={headers} pageName="somePageModifier" rows={rows} />)

            //test that the rendered row's cell has no contents, since return null when date is invalid
            expect(wrapper.find('.tableBody').childAt(0).children()).toHaveLength(0)
        })
    })
})
