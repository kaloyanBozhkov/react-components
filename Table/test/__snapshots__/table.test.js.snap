// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Table component rendering should render table component not configured properly msg, without any props  1`] = `
<p>
  Could not display table, please try again later
</p>
`;

exports[`testing Table component rendering should render table component when configured properly  1`] = `
<div
  className="table  someName"
>
  <header
    className="tableHeader"
  >
    <div
      key="0"
    >
      <p>
        columnHeader
      </p>
    </div>
  </header>
  <section
    className="tableBody"
    emptymessage="No data to show yet, try again later!"
  />
</div>
`;
